src/Symfony/Component/Serializer/Normalizer/ChainDenormalizer.php

```diff
@@ -0,0 +1,207 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\Serializer\Exception\LogicException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+use Symfony\Component\Serializer\Exception\PartialDenormalizationException;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * This holds a collection of denormalizers. It tries to be smart how it selects
+ * the correct Denormalizer.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class ChainDenormalizer implements DenormalizerInterface, SerializerAwareInterface
+{
+    private const SCALAR_TYPES = [
+        'int' => true,
+        'bool' => true,
+        'float' => true,
+        'string' => true,
+    ];
+
+    /**
+     * @deprecated since Symfony 7.1
+     */
+    private ?SerializerInterface $serializer = null;
+
+    /**
+     * @var DenormalizerInterface[]
+     */
+    private array $denormalizers = [];
+
+    /**
+     * @var array<string, array<string, array<bool>>>
+     */
+    private array $denormalizerCache = [];
+
+    /**
+     * @var array<string, array<class-string|'*'|'object'|string, bool|null>>
+     */
+    private array $supportedCache = [];
+
+    /**
+     * @param DenormalizerInterface[] $denormalizers
+     */
+    public function __construct(array $denormalizers = [])
+    {
+        foreach ($denormalizers as $denormalizer) {
+            $this->addDenormalizer($denormalizer);
+        }
+    }
+
+    public function addDenormalizer(DenormalizerInterface $denormalizer): void
+    {
+        if ($denormalizer instanceof DenormalizerAwareInterface) {
+            $denormalizer->setDenormalizer($this);
+        }
+
+        if (null !== $this->serializer && $denormalizer instanceof SerializerAwareInterface) {
+            $denormalizer->setSerializer($this->serializer);
+        }
+
+        $this->denormalizers[] = $denormalizer;
+        $this->denormalizerCache = [];
+        $this->supportedCache = [];
+    }
+
+    /**
+     * @throws NotNormalizableValueException
+     */
+    public function denormalize(mixed $data, string $type, ?string $format = null, array $context = []): mixed
+    {
+        if (isset($context[DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS], $context['not_normalizable_value_exceptions'])) {
+            throw new LogicException('Passing a value for "not_normalizable_value_exceptions" context key is not allowed.');
+        }
+
+        $denormalizer = $this->getDenormalizer($data, $type, $format, $context);
```

Comment:
There’s nothing to do here, I suppose, but it seems odd that we don’t first check if it’s a scalar value/type and skip this process earlier, knowing that denormalizers are intended to work with classes/objects only, or at least that’s what the `getDenormalizer` implementation suggests.

Reply:
Yes. You are very correct. 

This should be moved (in next PR) to a new `ScalarDenormalizer`. If we that in before 7.2 then there be no issues to keep BC.

Reply:
@yceruto you can have custom denormalizers for scalars if you want.

Reply:
Yeah, I was guessing that between the lines, but we should use better variable names to make it clearer for maintainers